miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/ParentFinder.java

```diff
@@ -0,0 +1,122 @@
+package uk.ac.bbsrc.tgac.miso.webapp.controller.rest;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import javax.ws.rs.core.Response.Status;
+
+import org.springframework.http.HttpEntity;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import uk.ac.bbsrc.tgac.miso.core.data.DetailedSample;
+import uk.ac.bbsrc.tgac.miso.core.data.Identifiable;
+import uk.ac.bbsrc.tgac.miso.core.data.Sample;
+import uk.ac.bbsrc.tgac.miso.core.util.LimsUtils;
+import uk.ac.bbsrc.tgac.miso.core.util.WhineyFunction;
+import uk.ac.bbsrc.tgac.miso.dto.Dtos;
+import uk.ac.bbsrc.tgac.miso.dto.SampleDto;
+
+public abstract class ParentFinder<M extends Identifiable> {
+  public static abstract class ParentAdapter<M extends Identifiable, P extends Identifiable, D> {
+
+    public abstract D asDto(P model);
+
+    public abstract P find(M model, Consumer<String> emitError);
+
+    private final String category;
+
+    public final String category() {
+      return category;
+    }
+
+    public ParentAdapter(String category) {
+      this.category = category;
+    }
+
+    public final HttpEntity<byte[]> handle(ObjectMapper mapper, Stream<M> items) throws JsonProcessingException {
+      List<String> errors = new ArrayList<>();
+      List<D> parents = items.map(item -> find(item, errors::add))//
+          .filter(Objects::nonNull)//
+          .collect(Collectors.groupingBy(Identifiable::getId)).values().stream()//
+          .map(l -> l.get(0))//
+          .map(this::asDto)//
+          .collect(Collectors.toList());
+      if (errors.size() > 0) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Use isEmpty() to check whether the collection is empty or not. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1155)